hnix-store-remote/src/System/Nix/Store/Remote.hs

```diff
@@ -32,3 +46,23 @@ verifyStore :: CheckFlag -> RepairFlag -> MonadStore Bool
 verifyStore check repair = simpleOpArgs VerifyStore $ do
   putBool check
   putBool repair
+
+addToStore :: forall hashAlgo. (NamedAlgo hashAlgo, ValidAlgo hashAlgo)
+    => StorePathName
+    -> FilePath
+    -> RecursiveFlag
+    -> PathFilter
+    -> RepairFlag
+    -> MonadStore LBS.ByteString
+addToStore name srcPath recursive filter repair = do
+  nar <- liftIO $ localPackNar narEffectsIO srcPath -- TODO actually use filter.
+  runOpArgs AddToStore $ do
+    putByteStringLen $ strToN $ unStorePathName name
+    putBool $ not (recursive && algoName @hashAlgo == "sha256") -- backwards compatibility hack
```

Comment:
Let's add an `algo :: HashAlgorithm` field to `ValidAlgo` instead?

Reply:
I tried that, but could not make it work, because then you need to define `algo` for Truncated, and you need to instantiate Eq for HashAlgorithm. I failed at both of these.

Reply:
Looking at this again, I think the fundamental problem is we're trying to pass the hash algo at the type level at all. Why not just take a value-level hash argument here? It's not like we construct a `Digest` or anything ourselves.

Reply:
Because we need a NamedHash to get the pass the corresponding algoName to the daemon.
How do you type a function that takes a HashAlgorith whose corresponding type is a a NamedHash instance ? And how do you access this instance from the value.

We could just take a string, but that would be kinda defeating the purpose of all of this.

Reply:
Argh, right. I'd say we should just have a value-level hashName function but then we'd want to exclude truncated hashes.

I'm torn between getting rid of this type level machinery altogether or separating out the base hash algorithms (sha, md5, etc.) from the digest types (base algo | truncated). Thoughts?

Reply:
Hmmm, I only ever programmed in haskell for the past "advent of code". I am therefore no expert in the habits of haskell programming.

But, if we want to mimic the remote store api as close as possible, then passing a real hash value as an argument would make sense. The current type annotation is strange.
Also, the Truncated stuff is weird to handle.
I was also surprised at how easy it is to create a "fake" hash. Turning the returned string into a correctly typed store path was very easy, but could be incorrect. A type system that can be easily escaped looks a bit like a "nullptr" in other languages.

May I propose to at least remove Truncated from the datatype ? It has no use outside of store paths while the other algos are used everywhere in nixpkgs.
That way we can make the resulting HashAlgorithms Eq. 

I guess we will get a better idea of what to do after implementing several api calls.

Reply:
OK, I will give this separation a shot.

Reply:
The further I think about this, the more it seems that using algoName could make sense. What does not make sense however, is how to create a valid StorePath from the string returned by the nix daemon.

The way I do it is not valid, as it creates a Digest from the base32-encoded hash. If you print a path generated like that, it re-encodes the digest to base32, making it too long.

https://github.com/haskell-nix/hnix-store/pull/53/files#diff-002378d5e26f547b360b3e048e2d2f9fR81-R88

If you split Truncated from the hash methods, a good use case is to implement ` makeStorePath :: BS.ByteString -> Maybe (StorePath storeDir)`